packages/medusa/src/services/cart.ts

```diff
@@ -1447,9 +1447,19 @@ class CartService extends TransactionBaseService {
             relations: ["rule", "rule.conditions", "regions"],
           })
 
-        await this.discountService_
-          .withTransaction(transactionManager)
-          .validateDiscountForCartOrThrow(cart, discounts)
+        try {
+          await this.discountService_
+            .withTransaction(transactionManager)
+            .validateDiscountForCartOrThrow(cart, discounts)
+        } catch (error) {
+          if (
+            error &&
+            error.type === MedusaError.Types.CUSTOMER_DISCOUNT_LIMIT_REACHED
+          ) {
+            await this.removeDiscount(cart.id, error.code)
+          }
+          throw error
+        }
```

Comment:
question: What case are you covering here? Customers are unable to add such discounts to the cart in the first place, so I would assume we don't need to remove it.

Reply:
@olivermrbl This is needed, because the cart can be created before we know which customer is connected to it. Therefore we added this logic to remove the discount when the error is related to a customer discount. 

Also maybe not a bad idea in general for Medusa to remove invalid discounts from carts, if due to cart updates discount rules are not valid anymore.